pkg/app/piped/appconfigreporter/appconfigreporter.go

```diff
@@ -0,0 +1,366 @@
+		}
+		apps = append(apps, appInfo)
+	}
+	return apps, nil
+}
+
+// updateUnregisteredApps sends all unregistered application configurations to the control-plane.
+func (r *Reporter) updateUnregisteredApps(ctx context.Context, registeredAppPaths map[string]struct{}) error {
+	apps := make([]*model.ApplicationInfo, 0)
+	for repoID, repo := range r.gitRepos {
+		as, err := r.findUnregisteredApps(ctx, repo.GetPath(), repoID, registeredAppPaths)
+		if err != nil {
+			return err
+		}
+		r.logger.Info(fmt.Sprintf("found out %d unregistered applications in repository %s", len(as), repoID))
+		apps = append(apps, as...)
+	}
+	if len(apps) == 0 {
+		return nil
+	}
+
+	_, err := r.apiClient.ReportUnregisteredApplicationConfigurations(
+		ctx,
+		&pipedservice.ReportUnregisteredApplicationConfigurationsRequest{
+			Applications: apps,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to put the latest unregistered application configurations: %w", err)
+	}
+	return nil
+}
+
+// findUnregisteredApps finds out unregistered application info in the given git repository.
+func (r *Reporter) findUnregisteredApps(ctx context.Context, repoPath, repoID string, registeredAppPaths map[string]struct{}) ([]*model.ApplicationInfo, error) {
+	return r.scanAllFiles(ctx, repoPath, repoID, registeredAppPaths, false)
+}
+
+// scanAllFiles inspects all files under the root or the given repository.
+// And gives back all application info as much as possible.
+func (r *Reporter) scanAllFiles(ctx context.Context, repoRoot, repoID string, registeredAppPaths map[string]struct{}, wantRegistered bool) ([]*model.ApplicationInfo, error) {
+	apps := make([]*model.ApplicationInfo, 0)
+	err := fs.WalkDir(r.fileSystem, repoRoot, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+
+		cfgRelPath, err := filepath.Rel(repoRoot, path)
+		if err != nil {
+			return err
+		}
+
+		if shouldSkip(repoID, cfgRelPath, registeredAppPaths, wantRegistered) {
+			return nil
+		}
+
+		appInfo, err := r.readApplicationInfo(ctx, path)
+		if err != nil {
+			r.logger.Error("failed to read application info",
+				zap.String("repo-id", repoID),
+				zap.String("config-file-path", cfgRelPath),
+				zap.Error(err),
+			)
+			return nil
+		}
+		apps = append(apps, appInfo)
+		// Continue reading so that it can return apps as much as possible.
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to inspect files under %s: %w", repoRoot, err)
+	}
+	return apps, nil
+}
+
+// cfgFilePath is a relative path from the repo root.
+func shouldSkip(repoID, cfgFilePath string, registeredAppPaths map[string]struct{}, wantRegistered bool) bool {
+	if !strings.HasSuffix(cfgFilePath, model.DefaultDeploymentConfigFileExtension) {
+		return true
+	}
+	gitPathKey := makeGitPathKey(repoID, cfgFilePath)
+	if _, registered := registeredAppPaths[gitPathKey]; registered != wantRegistered {
+		return true
+	}
+	return false
+}
```

Comment:
How about removing this function and using its code directly?
Because handling for both "registered" and "unregistered" cases makes it harder to read.

Reply:
Right. It should not be made common.